### DISCUSSION TITLE
Chaining CMPs

### DISCUSSION THREAD

**Discussion by rassie** [argoproj/argo-cd#10909] [General]
I would like to use [`kbld`](https://carvel.dev/kbld/) to pin images to their respective digests (useful for `latest` tags). 

Since `kbld` works as a filter, I would normally try to implement a minimal sidecar CMP. However I'm already using custom CMPs to generate my configuration, so that `kbld` would need to be running after my "normal" CMP is done and get generated configuration as input. To implement that I would either need to update my custom CMP to include `kbld` (which is more or less unnecessary mixing responsibilities) or "chain" multiple CMPs somehow, which is not a thing I'm aware of.

Is there any existing way or a concept how this could be done?

**Comment by crenshaw-dev** [comment]
I'd love to see this. I think it could technically be done today, but it would involve writing a gRPC client to manually 1) tar.gz a plugin's output, 2) invoke the other sidecar plugin's gRPC service (via its `.sock` file), and 3) extract the response and re-write it to standard out for the repo-server to consume.

A chainable-plugins enhancement would probably do precisely that, just providing native support via config flags.

An alternative would be chainable-sources. Once [multi-source plugins](https://github.com/argoproj/argo-cd/pull/10432) is merged, we could consider adding "other source" as a source-of-truth for a `sources` value instead of git or a Helm repo.

That was an absolute nightmare of a sentence, so here's an example, adapted from @ishitasequeira's example on the multi-source's PR:

```yaml
spec:
  sources:
    - repoURL: https://github.com/my-org/my-repo
      targetRevision: master
      ref: myRepo
    - from: "$myRepo"
      plugin:
        env:
          SOME: SETTING  
```

The repo-server would build the manifests from the first source, and then send them as a single file (say, manifests.yaml) to the second source for further processing.

Using `ref` would involve DAG resolution and cycle checking. We could instead just have sources evaluate sequentially.

**Comment by rassie** [comment]
Quite a bit food for thought, thank you.

Probably a rather naive question: is there any reason all of this couldn't be done (or at least prototyped) out-of-tree by something like a "controller" CMP, which would be the entrypoint and exit point and which would coordinate the chain of other CMPs?

**Reply by crenshaw-dev** [reply]
You could encapsulate the chaining in one "controller" plugin. But that controller would need a gRPC client to communicate with the other plugins, as well as mounts for the socket files. Definitely possible, but a bit involved.

I think the gRPC client could be exposed as part of the `argocd` CLI. The client already exists in the Argo CD code base (since it's used by the repo-server). It would just need to be exposed in a way that's accessible to plugins.
